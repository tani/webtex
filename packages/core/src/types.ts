interface LengthLike {
  _value: number;
  unit: string;
  add(other: LengthLike): LengthLike;
  sub(other: LengthLike): LengthLike;
  mul(scalar: number): LengthLike;
  div(scalar: number): LengthLike;
  norm(other: LengthLike): LengthLike;
}

interface LengthConstructor {
  new (value: number, unit: string): LengthLike;
  zero: LengthLike;
}

interface LengthGenerator {
  error(message: string): unknown;
  round(value: number): number;
  Length: LengthConstructor;
}

const makeLengthClass = (generator: LengthGenerator): LengthConstructor => {
  const g = generator;

  const unitsSp = new Map([
    ["sp", 1],
    ["pt", 65536],
    ["bp", (65536 * 72.27) / 72],
    ["pc", 65536 * 12],
    ["dd", (65536 * 1238) / 1157],
    ["cc", ((65536 * 1238) / 1157) * 12],
    ["in", 65536 * 72.27],
    ["px", (65536 * 72.27) / 96],
    ["mm", (65536 * 7227) / 2540],
    ["cm", (65536 * 7227) / 254],
  ]);

  class Length implements LengthLike {
    public _value: number = 0;
    public _unit: string = "";
    static zero: Length;

    constructor(value: number, unit: string) {
      if (typeof value !== "number") {
        g.error("Length CTOR: value needs to be a number!");
      }
      this._value = value;
      this._unit = unit;

      // Convert to scaled points if possible
      const unitValue = unitsSp.get(unit);
      if (unitValue !== undefined) {
        this._value = value * unitValue;
        this._unit = "sp";
      }
    }

    public toPx(): number {
      if (this._unit !== "sp") {
        g.error("Length.toPx() called on relative length!");
      }
      const pxValue = unitsSp.get("px");
      if (pxValue === undefined) {
        throw new Error("px unit conversion not found");
      }
      return g.round(this._value / pxValue);
    }

    public assertCompatible(other: Length, operation: string): void {
      if (this._unit !== other._unit) {
        g.error(
          `Length.${operation}: incompatible lengths! (${this._unit} and ${other._unit})`,
        );
      }
    }

    get value(): string {
      return this._unit === "sp"
        ? `${this.toPx()}px`
        : `${g.round(this._value)}${this._unit}`;
    }

    get px(): number {
      return this.toPx();
    }

    get pxpct(): string | number {
      return this._unit === "sp"
        ? this.toPx()
        : `${g.round(this._value)}${this._unit}`;
    }

    get unit(): string {
      return this._unit;
    }

    cmp(other: Length): number {
      this.assertCompatible(other, "cmp");
      return Math.sign(this._value - other._value);
    }

    add(other: Length): Length {
      this.assertCompatible(other, "add");
      return new Length(this._value + other._value, this._unit);
    }

    sub(other: Length): Length {
      this.assertCompatible(other, "sub");
      return new Length(this._value - other._value, this._unit);
    }

    mul(scalar: number): Length {
      return new Length(this._value * scalar, this._unit);
    }

    div(scalar: number): Length {
      return new Length(this._value / scalar, this._unit);
    }

    abs(): Length {
      return new Length(Math.abs(this._value), this._unit);
    }

    ratio(other: Length): number {
      this.assertCompatible(other, "ratio");
      return this._value / other._value;
    }

    norm(other: Length): Length {
      this.assertCompatible(other, "norm");
      return new Length(
        Math.sqrt(this._value ** 2 + other._value ** 2),
        this._unit,
      );
    }

    static min(...lengths: Length[]): Length {
      return lengths.reduce((a, b) => (a.cmp(b) < 0 ? a : b));
    }

    static max(...lengths: Length[]): Length {
      return lengths.reduce((a, b) => (a.cmp(b) > 0 ? a : b));
    }
  }

  Length.zero = new Length(0, "sp");
  return Length;
};

type ShiftDirection = "start" | "end";

interface ShiftCalc {
  x: number;
  y: number;
  msq: number;
  imsq: number;
  dir_x: number;
  dir_y: number;
}

class Vector<T extends LengthLike> {
  public _x: T;
  public _y: T;

  constructor(x: T, y: T) {
    this._x = x;
    this._y = y;
  }

  get x(): T {
    return this._x;
  }

  get y(): T {
    return this._y;
  }

  add(v: Vector<T>): Vector<T> {
    return new Vector(this._x.add(v.x) as T, this._y.add(v.y) as T);
  }

  sub(v: Vector<T>): Vector<T> {
    return new Vector(this._x.sub(v.x) as T, this._y.sub(v.y) as T);
  }

  mul(scalar: number): Vector<T> {
    return new Vector(this._x.mul(scalar) as T, this._y.mul(scalar) as T);
  }

  private assertCompatibleUnits(operation: string): void {
    if (this._x.unit !== this._y.unit) {
      throw new Error(
        `Vector.${operation}: incompatible lengths! (${this._x.unit} and ${this._y.unit})`,
      );
    }
  }

  private calculateShift(): ShiftCalc {
    const x = this._x._value;
    const y = this._y._value;

    return {
      x,
      y,
      msq: Math.sqrt(1 + (y * y) / (x * x)),
      imsq: Math.sqrt(1 + (x * x) / (y * y)),
      dir_x: x < 0 ? -1 : 1,
      dir_y: y < 0 ? -1 : 1,
    };
  }

  private performShift(l: T, direction: ShiftDirection): Vector<T> {
    this.assertCompatibleUnits(`shift_${direction}`);

    const calc = this.calculateShift();
    const { x, y, msq, imsq, dir_x, dir_y } = calc;

    const isStart = direction === "start";
    const multiplier = isStart ? -1 : 1;

    let newX: T;
    let newY: T;

    if (x !== 0 && y !== 0) {
      const shiftX = l.div(msq).mul(dir_x * multiplier) as T;
      const shiftY = l.div(imsq).mul(dir_y * multiplier) as T;
      newX = isStart ? shiftX : (this._x.add(shiftX) as T);
      newY = isStart ? shiftY : (this._y.add(shiftY) as T);
    } else if (y === 0) {
      const shiftX = l.mul(dir_x * multiplier) as T;
      newX = isStart ? shiftX : (this._x.add(shiftX) as T);
      newY = isStart ? (this._y.mul(0) as T) : this._y;
    } else {
      const shiftY = l.mul(dir_y * multiplier) as T;
      newX = isStart ? (this._x.mul(0) as T) : this._x;
      newY = isStart ? shiftY : (this._y.add(shiftY) as T);
    }

    return new Vector(newX, newY);
  }

  shift_start(l: T): Vector<T> {
    return this.performShift(l, "start");
  }

  shift_end(l: T): Vector<T> {
    return this.performShift(l, "end");
  }

  norm(): T {
    return this._x.norm(this._y) as T;
  }
}

export { makeLengthClass, Vector };
