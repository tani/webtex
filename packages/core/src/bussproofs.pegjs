// Bussproofs LaTeX Commands Grammar
// PEG.js grammar for parsing bussproofs package commands while preserving formula content as raw strings

start
  = bussproofs_command

bussproofs_command
  = axiom_command
  / unary_inf_command
  / binary_inf_command
  / trinary_inf_command
  / quaternary_inf_command
  / quinary_inf_command
  / abbreviated_command

// Core inference commands
axiom_command "axiom command"
  = "\\AxiomC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "axiom", 
        command: "AxiomC", 
        content: content 
      };
    }

unary_inf_command "unary inference command"
  = "\\UnaryInfC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "UnaryInfC", 
        arity: 1,
        content: content 
      };
    }

binary_inf_command "binary inference command"
  = "\\BinaryInfC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "BinaryInfC", 
        arity: 2,
        content: content 
      };
    }

trinary_inf_command "trinary inference command"
  = "\\TrinaryInfC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "TrinaryInfC", 
        arity: 3,
        content: content 
      };
    }

// Extended inference commands
quaternary_inf_command "quaternary inference command"
  = "\\QuaternaryInfC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "QuaternaryInfC", 
        arity: 4,
        content: content 
      };
    }

quinary_inf_command "quinary inference command"
  = "\\QuinaryInfC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "QuinaryInfC", 
        arity: 5,
        content: content 
      };
    }

// Abbreviated forms
abbreviated_command
  = axiom_abbreviated
  / unary_inf_abbreviated
  / binary_inf_abbreviated
  / trinary_inf_abbreviated

axiom_abbreviated "abbreviated axiom command"
  = "\\AXC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "axiom", 
        command: "AxiomC", 
        abbreviated: true,
        content: content 
      };
    }

unary_inf_abbreviated "abbreviated unary inference command"
  = "\\UIC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "UnaryInfC", 
        abbreviated: true,
        arity: 1,
        content: content 
      };
    }

binary_inf_abbreviated "abbreviated binary inference command"
  = "\\BIC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "BinaryInfC", 
        abbreviated: true,
        arity: 2,
        content: content 
      };
    }

trinary_inf_abbreviated "abbreviated trinary inference command"
  = "\\TIC" _ "{" content:raw_brace_content "}" {
      return { 
        type: "inference", 
        command: "TrinaryInfC", 
        abbreviated: true,
        arity: 3,
        content: content 
      };
    }

// Raw content extraction with proper brace matching
raw_brace_content
  = chars:(
      !("}" / "{") char:.  { return char; }
    / "{" nested:raw_brace_content "}" { return "{" + nested + "}"; }
    )* {
      return chars.join('');
    }

// Whitespace
_ "whitespace"
  = [ \t\n\r]*