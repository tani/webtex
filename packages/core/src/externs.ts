// Declarations for external globals (no runtime)
declare const _global: unknown;
declare const __e: unknown;
declare const __g: unknown;
