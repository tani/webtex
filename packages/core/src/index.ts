import he from "he";
import { Generator } from "./generator";
import { HtmlGenerator } from "./html-generator";
import { SyntaxError as ParserSyntaxError, parse } from "./latex-parser";
export { document, window } from "#window";
// Export types
export * from "./types";

export {
  he,
  parse,
  ParserSyntaxError as SyntaxError,
  Generator,
  HtmlGenerator,
};
