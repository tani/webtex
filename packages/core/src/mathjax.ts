import juice from "juice";

// MathJax type definitions
interface MathJaxConfig {
  loader: {
    source: Record<string, unknown>;
    require: (path: string) => Promise<unknown>;
    load: string[];
    paths: Record<string, string>;
  };
  tex: {
    packages: string[];
  };
  svg: {
    fontCache: string;
  };
  startup: {
    typeset: boolean;
    promise?: Promise<void>;
    adaptor?: {
      textContent: (node: unknown) => string;
      outerHTML: (node: unknown) => string;
    };
  };
  tex2svg?: (math: string, argv: { display: boolean }) => unknown;
  svgStylesheet?: () => unknown;
}

declare global {
  var MathJax: MathJaxConfig;
}

//
//  The default options
//
const importMap = {
  "mathjax/es5/adaptors/liteDOM.js": () =>
    import("mathjax/es5/adaptors/liteDOM.js"),
  "xyjax/build/xypic.js": () => import("xyjax/build/xypic.js"),
} as Record<string, () => Promise<unknown>>;

// Configure MathJax
globalThis.MathJax = {
  loader: {
    source: {},
    require: (path: string) => importMap[path](),
    load: ["adaptors/liteDOM", "[custom]/xypic"],
    paths: {
      mathjax: "mathjax/es5",
      custom: "xyjax/build",
    },
  },
  tex: {
    packages: [
      "base",
      "verb",
      "upgreek",
      "unicode",
      "textmacros",
      "textcomp",
      "tagformat",
      "setoptions",
      "require",
      "physics",
      "noundefined",
      "noerrors",
      "newcommand",
      "mhchem",
      "mathtools",
      "html",
      "gensymb",
      "extpfeil",
      "enclose",
      "empheq",
      "configmacros",
      "colorv2",
      "colortbl",
      "color",
      "centernot",
      "cases",
      "cancel",
      "bussproofs",
      "braket",
      "boldsymbol",
      "bbox",
      "autoload",
      "amscd",
      "ams",
      "action",
      "xyjax",
    ],
  },
  svg: {
    fontCache: "none",
  },
  startup: {
    typeset: false,
  },
};

//  Load the MathJax startup module
await import("mathjax/es5/tex-svg-full.js");

//  Wait for MathJax to start up
await MathJax.startup.promise;

export function tex2svg(math = "", argv = { display: true }) {
  if (!MathJax.tex2svg || !MathJax.startup.adaptor || !MathJax.svgStylesheet) {
    throw new Error("MathJax not properly initialized");
  }
  const node = MathJax.tex2svg(math, argv);
  const adaptor = MathJax.startup.adaptor;
  const stylesheet = adaptor.textContent(MathJax.svgStylesheet());
  const html = adaptor.outerHTML(node);
  return juice(`${html}<style>${stylesheet}</style>`);
}

// console.log(tex2svg("\\frac{a}{b}"));