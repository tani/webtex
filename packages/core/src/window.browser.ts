const window = globalThis.window;
const document = globalThis.document;

export { window, document };
