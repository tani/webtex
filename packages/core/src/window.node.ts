import { createHTMLWindow } from "svgdom";

const window = createHTMLWindow();
const document = window.document;

interface GlobalWithDOM {
  window: typeof window;
  document: typeof document;
}

const globalScope = globalThis as typeof globalThis & GlobalWithDOM;
globalScope.window = window;
globalScope.document = document;

export { window, document };
