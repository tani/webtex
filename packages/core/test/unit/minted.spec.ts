import { describe, expect, test } from "vitest";
import { HtmlGenerator, parse } from "../../src/index";

const renderDocument = (latex: string): string => {
  const generator = new HtmlGenerator({ hyphenate: false });
  const document = parse(latex, { generator }).htmlDocument();
  return document.documentElement.outerHTML;
};

describe("minted package", () => {
  test("renders highlighted code block", () => {
    const html = renderDocument(`\\documentclass{article}
\\usepackage{minted}
\\begin{document}
\\begin{minted}{javascript}
const answer = 42;
console.log(answer);
\\end{minted}
\\end{document}`);

    expect(html).toMatchSnapshot();
    expect(html).toContain("shiki");
    expect(html).toContain("minted-block");
  });

  test("applies line numbers when requested", () => {
    const html = renderDocument(`\\documentclass{article}
\\usepackage{minted}
\\begin{document}
\\begin{minted}[linenos,theme=github-light]{python}
def greet(name):
    print(f"Hello, {name}!")

if __name__ == "__main__":
    greet("WebTeX")
\\end{minted}
\\end{document}`);

    expect(html).toMatchSnapshot();
    expect(html).toContain("data-line-number");
    expect(html).toContain("minted-with-line-numbers");
  });
});
