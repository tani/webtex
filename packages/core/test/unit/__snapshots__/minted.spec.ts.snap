// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`minted package > applies line numbers when requested 1`] = `
"<html style="--size: 13.284px; --textwidth: 56.162%; --marginleftwidth: 21.919%; --marginrightwidth: 21.919%; --marginparwidth: 48.892%; --marginparsep: 14.612px; --marginparpush: 6.642px;"><head><title>untitled</title><meta charset="UTF-8"></meta><link type="text/css" rel="stylesheet" href="css/article.css"></head><body><div class="body" data-source-line="1" data-source-column="1"><pre class="shiki github-light minted-block minted-with-line-numbers" style="background-color:#fff;color:#24292e" tabindex="0" data-language="python"><code><span class="line" data-line-number="1"><span style="color:#D73A49">def</span><span style="color:#6F42C1"> greet</span><span style="color:#24292E">(name):</span></span>
<span class="line" data-line-number="2"><span style="color:#005CC5">    print</span><span style="color:#24292E">(</span><span style="color:#D73A49">f</span><span style="color:#032F62">&quot;Hello, </span><span style="color:#005CC5">{</span><span style="color:#24292E">name</span><span style="color:#005CC5">}</span><span style="color:#032F62">!&quot;</span><span style="color:#24292E">)</span></span>
<span class="line" data-line-number="3"></span>
<span class="line" data-line-number="4"><span style="color:#D73A49">if</span><span style="color:#005CC5"> __name__</span><span style="color:#D73A49"> ==</span><span style="color:#032F62"> &quot;__main__&quot;</span><span style="color:#24292E">:</span></span>
<span class="line" data-line-number="5"><span style="color:#24292E">    greet(</span><span style="color:#032F62">&quot;WebTeX&quot;</span><span style="color:#24292E">)</span></span></code></pre></div></body></html>"
`;

exports[`minted package > renders highlighted code block 1`] = `
"<html style="--size: 13.284px; --textwidth: 56.162%; --marginleftwidth: 21.919%; --marginrightwidth: 21.919%; --marginparwidth: 48.892%; --marginparsep: 14.612px; --marginparpush: 6.642px;"><head><title>untitled</title><meta charset="UTF-8"></meta><link type="text/css" rel="stylesheet" href="css/article.css"></head><body><div class="body" data-source-line="1" data-source-column="1"><pre class="shiki github-light minted-block" style="background-color:#fff;color:#24292e" tabindex="0" data-language="javascript"><code><span class="line"><span style="color:#D73A49">const</span><span style="color:#005CC5"> answer</span><span style="color:#D73A49"> =</span><span style="color:#005CC5"> 42</span><span style="color:#24292E">;</span></span>
<span class="line"><span style="color:#24292E">console.</span><span style="color:#6F42C1">log</span><span style="color:#24292E">(answer);</span></span></code></pre></div></body></html>"
`;
